### PR TITLE
bpo-45759: Better error messages for non-matching 'elif'/'else' state…

### PR DESCRIPTION
**Pull request by thatbirdguythatuknownot**
…ments (GH-29481)

https://bugs.python.org/issue45759

### CODE REVIEW

File: Grammar/python.gram [hunk 1]
```diff
@@ -128,6 +128,8 @@ simple_stmt[stmt_ty] (memo):
 compound_stmt[stmt_ty]:
     | &('def' | '@' | ASYNC) function_def
     | &'if' if_stmt
+    | &'elif' nomatch_elif
```
Comment (pablogsal) [inline Grammar/python.gram]: Unfortunately I'm not convinced about this, all of these should be in an "invalid_" rule and never inlined in the correct rule.

File: Grammar/python.gram [hunk 2]
```diff
@@ -358,6 +360,11 @@ else_block[asdl_stmt_seq*]:
     | invalid_else_stmt
     | 'else' &&':' b=block { b }
 
+nomatch_elif:
```
Comment (pablogsal) [inline Grammar/python.gram]: See my previous comment

File: Grammar/python.gram [hunk 3]
```diff
@@ -1221,20 +1230,32 @@ invalid_if_stmt:
...
+        RAISE_SYNTAX_ERROR_STARTING_FROM(a, "'elif' does not match a 'if' here. Maybe try unindenting?") }
+    | 'if' named_expression ':' (NEWLINE INDENT statements | simple_stmts) a='else' {
+        RAISE_SYNTAX_ERROR_STARTING_FROM(a, "'else' does not match a 'if'/'while'/'for'/'except' here. Maybe try unindenting?") }
 invalid_elif_stmt:
     | 'elif' named_expression NEWLINE { RAISE_SYNTAX_ERROR("expected ':'") }
     | a='elif' named_expression ':' NEWLINE !INDENT {
         RAISE_INDENTATION_ERROR("expected an indented block after 'elif' statement on line %d", a->lineno) }
+    | 'elif' named_expression ':' (NEWLINE INDENT statements | simple_stmts) a='elif' {
+        RAISE_SYNTAX_ERROR_STARTING_FROM(a, "'elif' does not match a 'if' here. Maybe try unindenting?") }
+    | 'elif' named_expression ':' (NEWLINE INDENT statements | simple_stmts) a='else' {
```
Comment (pablogsal) [inline Grammar/python.gram]: I an not particularly convinced that the message is generic regarding the match ('if'/'while'/'for'/'except')

**Review by pablogsal — CHANGES_REQUESTED** [review]
Thanks for your PR! We need to think about this a bit more because I am not fully sure if this error is common enough to justify the extra complexity.

File: Grammar/python.gram [hunk 4]
```diff
@@ -1221,20 +1230,32 @@ invalid_if_stmt:
     | 'if' named_expression NEWLINE { RAISE_SYNTAX_ERROR("expected ':'") }
     | a='if' a=named_expression ':' NEWLINE !INDENT {
         RAISE_INDENTATION_ERROR("expected an indented block after 'if' statement on line %d", a->lineno) }
+    | 'if' named_expression ':' (NEWLINE INDENT statements | simple_stmts) a='elif' {
```
Comment (pablogsal) [inline Grammar/python.gram]: You can probably merge this two

**Comment by thatbirdguythatuknownot** [comment]
Sorry for the many 'Fix conflicts' commits, I was trying to fix the `Parser/parser.c` conflict.

**Comment by thatbirdguythatuknownot** [comment]
I have made the requested changes; please review again.

**Comment by thatbirdguythatuknownot** [comment]
Moved to another pull request with same title.

**Pull request closed without merging**